core/node/eth_sender/src/eth_tx_manager.rs

```diff
@@ -549,7 +549,12 @@ impl EthTxManager {
             let l1_block_numbers = self.l1_interface.get_l1_block_numbers().await?;
             METRICS.track_block_numbers(&l1_block_numbers);
 
-            if last_known_l1_block < l1_block_numbers.latest {
+            // We need to re-run the iteration even if the latest seen block number is the same as the previous one, since
+            // in case of a low-activity settlement layer network, new blocks may only be produced by the transactions from the operator itself.
+            //
+            // We do not skip the check entirely, to ensure the monotonicity of the `last_known_l1_block`
+            // in case of bugs of the JSON-RPC API provider.
+            if last_known_l1_block <= l1_block_numbers.latest {
```

Comment:
I want to delete this check completely. 
https://github.com/matter-labs/zksync-era/pull/2535.

In this case, it doesn't make sense actually, the blocks most likely won't go down and if they will we have to reiterate on them.

Reply:
In this case it won't throttle anything

Reply:
You are right, the code needs to be changed so that if the block number didn't  change, then don't resent txs and only send new ones

Reply:
@tomg10 But don't we already have the following logic?

```
                if first_sent_at_block == Some(l1_block_numbers.latest.0) {
                    continue;
                }
```

Reply:
it works a bit different, if only prevents tx just sent at block N from being re-sent at block N. It won't prevent tx that was resent at block N not to be resent again in next iteration at block N

Reply:
we can probably modify it though

Reply:
so that it checks against last_sent_at_block, not first

Reply:
@tomg10 i've slightly changed the logic, could you verify it?

Reply:
I've just pushed a commit with a test for exactly that